File: android/app/src/main/java/com/dismoi/scout/Floating/Manager.java
package com.dismoi.scout.Floating;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.IBinder;
import android.text.Layout;

import com.dismoi.scout.Floating.FloatingService;
import com.dismoi.scout.Floating.OnCallback;

import com.dismoi.scout.Floating.Layout.Bubble;
import com.dismoi.scout.Floating.Layout.Message;

public class Manager {
  private static Manager INSTANCE;
  private final Context context;
  private boolean bounded;
  private FloatingService bubblesService;
  private int trashLayoutResourceId;
  private OnCallback listener;

  private static Manager getInstance(Context context) {
    if (INSTANCE == null) {
      INSTANCE = new Manager(context);
    }
    return INSTANCE;
  }

  private final ServiceConnection disMoiServiceConnection = new ServiceConnection() {
    @Override
    public void onServiceConnected(ComponentName name, IBinder service) {
      FloatingService.FloatingServiceBinder binder = (FloatingService.FloatingServiceBinder)service;
      Manager.this.bubblesService = binder.getService();
      configureBubblesService();
      bounded = true;
      if (listener != null) {
        listener.onInitialized();
      }
    }

    @Override
    public void onServiceDisconnected(ComponentName name) {
      bounded = false;
    }
  };

  private Manager(Context context) {
    this.context = context;
  }

  private void configureBubblesService() {
    bubblesService.addTrash(trashLayoutResourceId);
  }

  public void initialize() {
    context.bindService(new Intent(context, FloatingService.class),
            disMoiServiceConnection,
            Context.BIND_AUTO_CREATE);
  }

  public void recycle() {
    context.unbindService(disMoiServiceConnection);
  }

  public void addDisMoiBubble(Bubble bubble, int x, int y) {
    if (bounded) {
      bubblesService.addDisMoiBubble(bubble, x, y);
    }
  }

  public void addDisMoiMessage(Message message, int x, int y) {
    if (bounded) {
      bubblesService.addDisMoiMessage(message, x, y);
    }
  }

  public void removeBubble(Bubble bubble) {
    if (bounded) {
      bubblesService.removeBubble(bubble);
    }
  }

  public void removeDisMoiMessage(Message message) {
    if (bounded) {
      bubblesService.removeMessage(message);
    }
  }

  public static class Builder {
    private final Manager disMoiManager;

    public Builder(Context context) {
      this.disMoiManager = getInstance(context);
    }

    public Builder setInitializationCallback(OnCallback listener) {
      disMoiManager.listener = listener;
      return this;
    }

    public Builder setTrashLayout(int trashLayoutResourceId) {
      disMoiManager.trashLayoutResourceId = trashLayoutResourceId;
      return this;
    }

    public Manager build() {
      return disMoiManager;
    }
  }
}

File: android/app/src/main/java/com/dismoi/scout/floating/Manager.kt
package com.dismoi.scout.floating

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.IBinder
import com.dismoi.scout.floating.FloatingService.FloatingServiceBinder
import com.dismoi.scout.floating.layout.Bubble
import com.dismoi.scout.floating.layout.Message

class Manager private constructor(private val context: Context) {
  private var bounded = false
  private var bubblesService: FloatingService? = null
  private var trashLayoutResourceId = 0
  private var listener: OnCallback? = null
  private val disMoiServiceConnection: ServiceConnection = object : ServiceConnection {
    override fun onServiceConnected(name: ComponentName, service: IBinder) {
      val binder = service as FloatingServiceBinder
      bubblesService = binder.service
      configureBubblesService()
      bounded = true
      if (listener != null) {
        listener!!.onInitialized()
      }
    }

    override fun onServiceDisconnected(name: ComponentName) {
      bounded = false
    }
  }

  private fun configureBubblesService() {
    bubblesService!!.addTrash(trashLayoutResourceId)
  }

  fun initialize() {
    context.bindService(
      Intent(context, FloatingService::class.java),
      disMoiServiceConnection,
      Context.BIND_AUTO_CREATE
    )
  }

  fun recycle() {
    context.unbindService(disMoiServiceConnection)
  }

  fun addDisMoiBubble(bubble: Bubble?, x: Int, y: Int) {
    if (bounded) {
      bubblesService!!.addDisMoiBubble(bubble!!, x, y)
    }
  }

  fun addDisMoiMessage(message: Message?, x: Int, y: Int) {
    if (bounded) {
      bubblesService!!.addDisMoiMessage(message, x, y)
    }
  }

  fun removeBubble(bubble: Bubble?) {
    if (bounded) {
      bubblesService!!.removeBubble(bubble!!)
    }
  }

  fun removeDisMoiMessage(message: Message?) {
    if (bounded) {
      bubblesService!!.removeMessage(message)
    }
  }

  class Builder(context: Context) {
    private val disMoiManager: Manager
    fun setInitializationCallback(listener: OnCallback?): Builder {
      disMoiManager.listener = listener
      return this
    }

    fun setTrashLayout(trashLayoutResourceId: Int): Builder {
      disMoiManager.trashLayoutResourceId = trashLayoutResourceId
      return this
    }

    fun build(): Manager {
      return disMoiManager
    }

    init {
      disMoiManager = getInstance(context)
    }
  }

  companion object {
    private fun getInstance(context: Context): Manager {
      return Manager(context)
    }
  }
}
